make/devices.cmake

```diff
@@ -0,0 +1,754 @@
+      message(FATAL_ERROR "Required argument ${ARG} is the empty string.")
+    endif()
+    set_target_properties(
+      ${DEFINE_ARCH_ARCH}
+      PROPERTIES ${ARG} "${DEFINE_ARCH_${ARG}}"
+    )
+  endforeach()
+endfunction()
+
+function(DEFINE_DEVICE_TYPE)
+  # ~~~
+  # DEFINE_DEVICE_TYPE(
+  #   DEVICE_TYPE <device_type>
+  #   ARCH <arch>
+  #   ARCH_XML <arch.xml>
+  #   )
+  # ~~~
+  #
+  # Defines a device type with the specified architecture.  ARCH_XML argument
+  # must be a file target (see MAKE_FILE_TARGET).
+  #
+  # DEFINE_DEVICE_TYPE defines a dummy target <arch>_<device_type>_arch that
+  # will build the merged architecture file for the device type.
+  set(options)
+  set(oneValueArgs DEVICE_TYPE ARCH ARCH_XML)
+  set(multiValueArgs)
+  cmake_parse_arguments(
+    DEFINE_DEVICE_TYPE
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+  )
+
+  add_custom_target(${DEFINE_DEVICE_TYPE_DEVICE_TYPE})
+  foreach(ARG ARCH)
+    if("${DEFINE_DEVICE_TYPE_${ARG}}" STREQUAL "")
+      message(FATAL_ERROR "Required argument ${ARG} is the empty string.")
+    endif()
+    set_target_properties(
+      ${DEFINE_DEVICE_TYPE_DEVICE_TYPE}
+      PROPERTIES ${ARG} ${DEFINE_DEVICE_TYPE_${ARG}}
+    )
+  endforeach()
+
+  #
+  # Generate a arch.xml for a device.
+  #
+  set(DEVICE_MERGED_FILE arch.merged.xml)
+
+  set(MERGE_XML_XSL ${symbiflow-arch-defs_SOURCE_DIR}/common/xml/xmlsort.xsl)
+  set(
+    MERGE_XML_INPUT ${CMAKE_CURRENT_BINARY_DIR}/${DEFINE_DEVICE_TYPE_ARCH_XML}
+  )
+  get_file_target(MERGE_XML_INPUT_TARGET ${DEFINE_DEVICE_TYPE_ARCH_XML})
+  set(MERGE_XML_OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${DEVICE_MERGED_FILE})
+
+  get_target_property_required(XSLTPROC env XSLTPROC)
+  add_custom_command(
+    OUTPUT ${MERGE_XML_OUTPUT}
+    DEPENDS ${MERGE_XML_XSL} ${MERGE_XML_INPUT} ${MERGE_XML_INPUT_TARGET}
+    COMMAND
+      ${CMAKE_COMMAND} -E make_directory
+      ${CMAKE_CURRENT_BINARY_DIR}/${OUT_DEVICE_DIR}
+    COMMAND
+      ${XSLTPROC}
+      --nomkdir
+      --nonet
+      --xinclude
+      --output ${MERGE_XML_OUTPUT} ${MERGE_XML_XSL} ${MERGE_XML_INPUT}
+  )
+  add_custom_target(
+    ${DEFINE_DEVICE_TYPE_ARCH}_${DEFINE_DEVICE_TYPE_DEVICE_TYPE}_arch
+    DEPENDS ${DEVICE_MERGED_FILE}
+  )
+
+  make_file_target(FILE ${DEVICE_MERGED_FILE} GENERATED)
+
+  set_target_properties(
+    ${DEFINE_DEVICE_TYPE_DEVICE_TYPE}
+    PROPERTIES
+      DEVICE_MERGED_FILE ${CMAKE_CURRENT_SOURCE_DIR}/${DEVICE_MERGED_FILE}
+  )
+
+endfunction()
+
+function(DEFINE_DEVICE)
+  # ~~~
+  # DEFINE_DEVICE(
+  #   DEVICE <device>
+  #   ARCH <arch>
+  #   DEVICE_TYPE <device_type>
+  #   PACKAGES <list of packages>
+  #   )
+  # ~~~
+  #
+  # Defines a device within a specified FPGA architecture.
+  #
+  # Creates dummy targets <arch>_<device>_<package>_rrxml_virt and
+  # <arch>_<device>_<package>_rrxml_virt  that generates the the virtual and
+  # real rr_graph for a specific device and package.
```

Comment:
FYI - We want to get rid of the `-virt` in the device names sometime -> https://github.com/SymbiFlow/symbiflow-arch-defs/issues/177

Reply:
Ack